# Извлечение навыков из поля `description`

## Импорт библиотек

In [51]:
import re
import pandas as pd

from pathlib import Path
from itertools import chain

In [52]:
# Хранилище обработанных данных:

PROCESSED_DIR = Path('../data/processed')
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)


## Работа с данными

In [9]:
df = pd.read_csv("https://izemenkov.github.io/skillgrow/data/raw_data/vacancies_master.csv")

# Оставляем только описание
texts = df["description"].dropna().copy()

# Простейшая очистка
def clean_text(t):
    t = re.sub(r"<.*?>", " ", t)         # убираем html разметку, если сохранилась
    t = re.sub(r"[^а-яА-Яa-zA-Z0-9+#/\. ]", " ", t)  # убираем спецсимволы
    t = re.sub(r"\s+", " ", t)
    return t.lower().strip()

texts = texts.apply(clean_text)

df["clean_text"] = texts
df.head()

id                                         name             area  \
0  127202494                 ML-аналитик / Data Scientist           Москва   
1  127244399                 Аналитик данных/Data Analyst  Нижний Новгород   
2  126209623        Стажер, Data Analyst / Data Scientist           Москва   
3  127192646  Младший аналитик данных/Junior Data Analyst           Москва   
4  126562421          Junior Data Analyst/Аналитик данных           Казань   

           experience                                         key_skills  \
0           Нет опыта                                                NaN   
1           Нет опыта            Python, SQL, Базы данных, Анализ данных   
2           Нет опыта  Python, pandas, Numpy, ООП, Алгоритмы и структ...   
3  От 1 года до 3 лет  Python, SQL, Аналитическое мышление, Исследова...   
4           Нет опыта                                                NaN   

                                         description  \
0  Обязанности:  Моделирование клиентского поведе...   
1  Ищем специалиста, готового влиться в команду п...   
2  В сегодняшней бизнес-среде компаниям необходим...   
3  Mediascope – исследовательская компания, котор...   
4  Привет! Мы команда разработчиков платформы You...   

                                          clean_text  
0  обязанности моделирование клиентского поведени...  
1  ищем специалиста готового влиться в команду пр...  
2  в сегодняшней бизнес среде компаниям необходим...  
3  mediascope исследовательская компания которая ...  
4  привет мы команда разработчиков платформы youn...

In [13]:
# Создадим список навыков
df["skills_list"] = df["key_skills"].fillna("").apply(lambda x: [s.strip().lower() for s in x.split(",") if s.strip()])

# объединяем все навыки в один список
all_skills = list(chain.from_iterable(df["skills_list"]))

skill_freq = (
    pd.Series(all_skills)
    .value_counts()
    .reset_index()
    .rename(columns={"index": "skill", 0: "count"})
)

skill_freq.head(20)


skill  count
0                      python    190
1                         sql    140
2                      pandas     44
3                     pytorch     37
4                          ml     36
5                    big data     36
6               анализ данных     34
7   математическая статистика     31
8                       numpy     24
9                      hadoop     23
10                   power bi     22
11        визуализация данных     22
12                        nlp     21
13               data science     20
14           machine learning     19
15                 tensorflow     18
16              deep learning     18
17                 clickhouse     18
18                        git     17
19               scikit-learn     16

In [17]:
# Проверим на "шум"

# Навыки короче 2 символов
skill_freq[skill_freq.skill.str.len() < 2]

# Видим только r, оставим

skill  count
58     r      5

In [20]:
# Навыки, встречающиеся менее 3 раз
skill_freq[skill_freq["count"] < 3]

skill  count
90                     svd      2
91   математический анализ      2
92                 дашборд      2
93                datalens      2
94                     rag      2
..                     ...    ...
313         argo workflows      1
314                    с++      1
315   разработка платформы      1
316                  tdorp      1
317                    ооп      1

[228 rows x 2 columns]

Навыки, встречающиеся менее двух раз уберем. Среди них, визуально, есть и значащие, но, зачастую, они являются подразделами более частых навыков. Попробуем обойтись без них

In [28]:
# убираем навыки частотой < 2 и навык it, как незначащий
skill_vocab = skill_freq[(skill_freq["count"] >= 3) & (skill_freq["skill"] != 'it')]["skill"].tolist()
len(skill_vocab), skill_vocab[:30]


(89,
 ['python',
  'sql',
  'pandas',
  'pytorch',
  'ml',
  'big data',
  'анализ данных',
  'математическая статистика',
  'numpy',
  'hadoop',
  'power bi',
  'визуализация данных',
  'nlp',
  'data science',
  'machine learning',
  'tensorflow',
  'deep learning',
  'clickhouse',
  'git',
  'scikit-learn',
  'mlflow',
  'postgresql',
  'docker',
  'ms excel',
  'pyspark',
  'data analysis',
  'etl',
  'разработка поисковых технологий',
  'a/b тесты',
  'llm'])

In [29]:
# Извлекаем навыки с помощью полученного списка:

df["clean_text"] = df["description"].fillna("").str.lower()

def extract_from_text(text, vocab):
    results = []
    for s in vocab:
        if len(s) == 1:
            # skill состоит из 1 буквы: ищем только как отдельное слово
            # пример: r → ищем ` r ` или начало/конец строки
            pattern = r'\b' + re.escape(s) + r'\b'
            if re.search(pattern, text):
                results.append(s)

        else:
            # обычные навыки: разрешаем подстроки
            if s in text:
                results.append(s)

    return results

df["skills_from_description"] = df["clean_text"].apply(lambda t: extract_from_text(t, skill_vocab))


In [30]:
# Проверим покрытие
coverage = (df["skills_from_description"].apply(len) > 0).mean()
print(f"Доля вакансий с навыками из описания: {coverage:.1%}")

skills_flat = list(chain.from_iterable(df["skills_from_description"]))
pd.Series(skills_flat).value_counts().head(20)


Доля вакансий с навыками из описания: 99.6%


python          391
sql             318
ml              304
pandas          175
pytorch         141
numpy           134
llm             130
nlp             119
аналитика       118
git             115
spark           110
docker          100
api              88
scikit-learn     72
data science     70
clickhouse       65
power bi         65
pyspark          64
hadoop           62
etl              62
Name: count, dtype: int64

Удалось добиться хорошего покрытия вакансий навыками.
Теперь приедем все к единому полю навыков, с которым будем работать в дальнейшем

In [35]:
def unite_skills(row):
    s1 = row.get("skills_list", [])
    s2 = row.get("skills_from_description", [])
    # объединяем, чистим и сортируем
    return sorted(set(map(str.lower, s1 + s2)))

df["final_skills"] = df.apply(unite_skills, axis=1)


In [41]:
# Смотрим частоты по финальному списку навыков

final_skills_flat = list(chain.from_iterable(df["final_skills"]))
pd.Series(final_skills_flat).value_counts().head(25)


python           407
sql              333
ml               306
pandas           186
pytorch          150
numpy            143
llm              131
git              127
аналитика        119
nlp              119
spark            115
docker           105
api               88
data science      82
scikit-learn      79
hadoop            76
clickhouse        71
pyspark           71
power bi          71
анализ данных     68
big data          64
tensorflow        62
etl               62
postgresql        60
matplotlib        60
Name: count, dtype: int64

In [43]:
# Смотрим частотность навыков, исходя из опыта
exploded = df.explode("final_skills")
grp = exploded.groupby(["experience", "final_skills"]).size().reset_index(name="count")
grp.sort_values("count", ascending=False).groupby("experience").head(15)


experience               final_skills  count
432       От 3 до 6 лет                     python    218
451       От 3 до 6 лет                        sql    173
402       От 3 до 6 лет                         ml    173
226  От 1 года до 3 лет                     python    137
237  От 1 года до 3 лет                        sql    115
433       От 3 до 6 лет                    pytorch     94
196  От 1 года до 3 лет                         ml     91
421       От 3 до 6 лет                     pandas     90
397       От 3 до 6 лет                        llm     78
415       От 3 до 6 лет                      numpy     73
413       От 3 до 6 лет                        nlp     72
366       От 3 до 6 лет                     docker     64
218  От 1 года до 3 лет                     pandas     64
373       От 3 до 6 лет                        git     63
450       От 3 до 6 лет                      spark     62
476       От 3 до 6 лет                  аналитика     56
260  От 1 года до 3 лет                  аналитика     54
211  От 1 года до 3 лет                      numpy     52
346       От 3 до 6 лет                        api     48
379       От 3 до 6 лет                     hadoop     47
173  От 1 года до 3 лет                        git     44
430       От 3 до 6 лет                    pyspark     42
191  От 1 года до 3 лет                        llm     42
227  От 1 года до 3 лет                    pytorch     40
140  От 1 года до 3 лет                        api     38
208  От 1 года до 3 лет                        nlp     36
40          Более 6 лет                         ml     34
236  От 1 года до 3 лет                      spark     34
257  От 1 года до 3 лет              анализ данных     34
54          Более 6 лет                     python     33
167  От 1 года до 3 лет                     docker     33
64          Более 6 лет                        sql     30
195  От 1 года до 3 лет                 matplotlib     30
49          Более 6 лет                     pandas     21
113           Нет опыта                     python     19
23          Более 6 лет                        git     18
63          Более 6 лет                      spark     15
119           Нет опыта                        sql     15
7           Более 6 лет                   big data     14
55          Более 6 лет                    pytorch     14
12          Более 6 лет               data science     13
52          Более 6 лет                   power bi     12
36          Более 6 лет                        llm     11
18          Более 6 лет                        dwh     11
45          Более 6 лет                        nlp     11
11          Более 6 лет                         cv     11
19          Более 6 лет                        etl     11
109           Нет опыта                     pandas     11
103           Нет опыта                         ml      8
107           Нет опыта                      numpy      8
83            Нет опыта                   big data      6
111           Нет опыта                   power bi      6
126           Нет опыта                  аналитика      5
130           Нет опыта  математическая статистика      5
112           Нет опыта                    pyspark      4
120           Нет опыта                    tableau      4
118           Нет опыта                      spark      4
101           Нет опыта           machine learning      4
125           Нет опыта              анализ данных      4
89            Нет опыта               data science      4

### Выводы

* Удалось добиться извлечения навыков из вакансий с хорошим покрытием (>99%)
* Собрали ядро DS компетенций для рынка РФ
| Навык                                  | Частота                                                           | Комментарий                             |
| -------------------------------------- | ----------------------------------------------------------------- | --------------------------------------- |
| **python**                             | 407                                                               | базовый рабочий язык                    |
| **sql**                                | 333                                                               | must-have для аналитики и продакшена    |
| **ml**                                 | 306                                                               | обобщающее, но отражает роль            |
| **pandas / numpy**                     | 186 / 143                                                         | работа с табличными данными             |
| **pytorch / tensorflow**               | 150 / 62                                                          | модели в проде, особенно pytorch        |
| **nlp / llm**                          | 119 / 131                                                         | **всплеск из-за LLM тренда**, это важно |
| **spark / hadoop / pyspark**           | 115 / 76 / 71                                                     | Big Data → всё чаще must-have           |
| **docker / git / api**                 | 105 / 127 / 88                                                    | продакшен навыки                        |
| **power bi / clickhouse / postgresql** | Power BI = BI для бизнеса; ClickHouse/PostgreSQL                  | инфраструктура                       |


* Навыки против опыта показывают возможный карьерный трек специалиста:

| Опыт          | Ключевые навыки                                              | Интерпретация                                |
| ------------- | ------------------------------------------------------------ | -------------------------------------------- |
| **Нет опыта** | python, sql, pandas, power bi                                | Вход с BI / аналитики ↔ стажировки           |
| **1–3 года**  | python, sql, ml, pandas, nlp, numpy, docker                  | переход «анализ → модели → прототипирование» |
| **3–6 лет**   | python, sql, llm, pytorch, spark, docker, clickhouse, hadoop | продакшн ML + Big Data + архитектура         |
| **> 6 лет**   | python, sql, ml, big data, power bi, cv                      | роли "ведущий / архитектор / руководитель"   |


MVP Системы рекомендаций может работать следующим образом:

"Вы целитесь в позицию с опытом X. Средние навыки для этой зоны: ...
У вас есть: ...
Рекомендуем добрать: ..."


In [54]:
# Сохраним файл мастер-данных в ../data/processed
df_to_save = df[['id','name','area','experience','description','final_skills']]

In [55]:
df_to_save

id                                           name  \
0    127202494                   ML-аналитик / Data Scientist   
1    127244399                   Аналитик данных/Data Analyst   
2    126209623          Стажер, Data Analyst / Data Scientist   
3    127192646    Младший аналитик данных/Junior Data Analyst   
4    126562421            Junior Data Analyst/Аналитик данных   
..         ...                                            ...   
473  126190799      ML-разработчик (рекомендательные системы)   
474  126637726  Старший ML-разработчик в Яндекс Плюс и AdTech   
475  126638664      Специалист по машинному обучению (middle)   
476  126638661      Специалист по машинному обучению (middle)   
477  126018299      ML-разработчик (рекомендательные системы)   

                area          experience  \
0             Москва           Нет опыта   
1    Нижний Новгород           Нет опыта   
2             Москва           Нет опыта   
3             Москва  От 1 года до 3 лет   
4             Казань           Нет опыта   
..               ...                 ...   
473  Санкт-Петербург       От 3 до 6 лет   
474           Москва       От 3 до 6 лет   
475  Нижний Новгород       От 3 до 6 лет   
476           Самара       От 3 до 6 лет   
477           Москва       От 3 до 6 лет   

                                           description  \
0    Обязанности:  Моделирование клиентского поведе...   
1    Ищем специалиста, готового влиться в команду п...   
2    В сегодняшней бизнес-среде компаниям необходим...   
3    Mediascope – исследовательская компания, котор...   
4    Привет! Мы команда разработчиков платформы You...   
..                                                 ...   
473  Дзен — сложная рекомендательная система и блог...   
474  Яндекс Плюс — это единая подписка на сервисы Я...   
475  Чем предстоит заниматься  Cоздавать, внедрять ...   
476  Чем предстоит заниматься  Cоздавать, внедрять ...   
477  Дзен — сложная рекомендательная система и блог...   

                                          final_skills  
0    [etl, ml, mlflow, pandas, python, scikit-learn...  
1    [dwh, python, sql, анализ данных, аналитика, б...  
2    [big data, data science, git, ml, ms sql, nump...  
3    [data science, dwh, jupyter notebook, python, ...  
4      [numpy, pandas, power bi, python, sql, tableau]  
..                                                 ...  
473  [catboost, java, lightgbm, ml, nlp, recsys, sv...  
474  [java, ml, python, sql, алгоритмы и структуры ...  
475  [git, linux, llm, ml, nlp, pyspark, python, r,...  
476  [git, linux, llm, ml, nlp, pyspark, python, r,...  
477  [catboost, java, lightgbm, ml, nlp, recsys, sv...  

[478 rows x 6 columns]

In [56]:
df_to_save.to_csv(PROCESSED_DIR / 'extracted_skills.csv', index=False)